### PR TITLE
[java] DuplicateImports reported for the same import... and import static... #2546

### LINKED ISSUE #2546: [java] DuplicateImports reported for the same import... and import static...
**Affects PMD Version:** [6.24.0](https://pmd.github.io/latest/pmd_release_notes.html#24-may-2020---6240)

**Rule:** [DuplicateImports](https://pmd.github.io/latest/pmd_rules_java_codestyle.html#duplicateimports)

**Description:**

Violation of DuplicateImports reported for (star) imports of types and statics from other class.

**Code Sample demonstrating the issue:**
https://github.com/pzrep/pmd-DuplicateImports

```
import charles.X.*;
import static charles.X.*;
```
is reported as violation, although they import different things, and one does not replace another.

**Expected outcome:**

No violation of *DuplicateImports* reported.

*   Does PMD report a violation, where there shouldn't be one? -> false-positive

I think so.

<sup>*   Is PMD missing to report a violation, where there should be one? -> false-negative
No.</sup>

**Running PMD through:** CLI in sample, found with maven.

### PR DESCRIPTION
**Pull request by zyc-Iroha**
## Describe the PR

Modify
pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/codestyle/DuplicateImportsRule.java
pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/ImportWrapper.java
to avoid detecting DuplicateImports in issue #2546

## Related issues

- Fixes #2546 

## Ready?

- [ ] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Complete build `./mvnw clean verify` passes (checked automatically by travis)
- [x] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">This changeset introduces 0 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/b64ea1d95fc1412aa4b3101e58122a0f/diff/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/ImportWrapper.java [hunk 1]
```diff
@@ -93,7 +93,7 @@ public int hashCode() {
         if (name == null) {
             return fullname.hashCode();
         }
-        return name.hashCode();
+        return name.hashCode() + (isStaticDemand ? 1 : 0);
```
Comment (adangel) [inline pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/ImportWrapper.java:96]: When we change hashCode, we probably need to change equals as well.
With your change now, two instances of ImportWrapper could be equal (having the same name) but still give different hashCodes. That violates the general contract of hashCode/equals... https://docs.oracle.com/en/java/javase/14/docs/api/java.base/java/lang/Object.html#hashCode()

> If two objects are equal according to the equals(Object) method, then calling the hashCode method on each of the two objects must produce the same integer result. 

I guess, we first need to make sure, that wrapped "import" and "import static" are not equal. They may optionally have a different hash code (but that's just an optimization and not required). It is allowed, that two different (unequal) objects return the same hashCode (but it's not efficient - e.g. if put many of such objects into a hashmap, they would end up all in the same bucket and creating many collisions).

With your change of the hashCode function, you are relying on the implementation detail of "HashSet", that it actually uses this hashCode function.... which could change.

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/codestyle/DuplicateImportsRule.java [hunk 2]
```diff
@@ -88,22 +88,31 @@ private boolean isDisambiguationImport(ASTCompilationUnit node, String singleTyp
...
         ImportWrapper wrapper = new ImportWrapper(node.getImportedName(), node.getImportedName(),
                 node, node.isStatic() && node.isImportOnDemand());
-
-        // blahhhh... this really wants to be ASTImportDeclaration to be
-        // polymorphic...
+        boolean isViolation = false;
         if (node.isImportOnDemand()) {
             if (importOnDemandImports.contains(wrapper)) {
-                addViolation(data, node, node.getImportedName());
+                for (ImportWrapper thisImportOnDemand : importOnDemandImports) {
```
Comment (adangel) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/codestyle/DuplicateImportsRule.java:94]: Maybe we don't need the changes here at all, if we correct the equals implementation of ImportWrapper.
A Unit/Rule test should tell us...

**Comment by adangel** [comment]
Replaced by #2644

**Pull request closed without merging**